fendermint/vm/topdown/src/voting.rs

```diff
@@ -0,0 +1,249 @@
+// Copyright 2022-2024 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use async_stm::{abort, retry, Stm, StmResult, TVar};
+use std::fmt::Debug;
+use std::hash::Hash;
+
+use crate::{BlockHash, BlockHeight};
+
+// Usign this type because it's `Hash`, unlike the normal `libsecp256k1::PublicKey`.
+use ipc_ipld_resolver::ValidatorKey;
+
+pub type Weight = u64;
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error<K = ValidatorKey, V: AsRef<[u8]> = BlockHash> {
+    #[error("failed to extend chain; expected block height {0}, got {1}")]
+    UnexpectedBlock(BlockHeight, BlockHeight),
+
+    #[error("unknown validator: {0:?}")]
+    UnknownValidator(K),
+
+    #[error(
+        "equivocation by validator {0:?} at height {1}; {} != {}",
+        hex::encode(.2),
+        hex::encode(.3)
+    )]
+    Equivocation(K, BlockHeight, V, V),
```

Comment:
Oh, no worries, I can explain. I tried to quickly find; here's one in the CometBFT docs about [Evidence](https://docs.cometbft.com/v0.38/spec/consensus/evidence#equivocation).

It's (what a I think is a common) term in Proof-of-Stake consensus when a Byzantine validator supports two contradicting facts at the same time, where an honest validator can only support one. In this instance we are trying to find a finalized chain, which should be evident if one observes the parent chain; if a validator casts votes for two different block hashes on the same height in the parent chain (on the _finalized_ chain, not the tip where there is forking in the Longest Chain Filecoin), then they are not being honest, or they have misconfigured their node and their finality threshold is too low.

In any case we should ignore their votes because if everyone was allowed to behave this way we'd end up finalizing something random. Hopefully no more than 1/3rd of the voting power ever casts contradicting votes, and then we can still finalize blocks. 

It's something I put in, but didn't pursue to the end where if you equivocate at all then you are out: you can equivocate on height 4, but not on 5, and that will still count as legal on 5. Or you can support chain A on height 4, chain B on height 5, then chain A on height 6 again, and I won't see it as equivocation even when I know that 4 and 6 formed a chain. The potential is there, but currently it's only used for the most obvious case where on the same height there are two conflicting votes from the same validator.

Reply:
Oh nice!, yeah I think I figured out what this would be used for when I read rest of the PR, but thanks about the details here, super useful!